### PR TITLE
Move UriSigner from HttpKernel to HttpFoundation package

### LINKED ISSUE #50384: [HttpKernel] Move UriSigner to Http Foundation
### Description

Based on [Symfony Docs](https://github.com/symfony/symfony-docs/issues/18330) discussion. I want to open a discussion here to maybe move the `UriSigner` class to the `http-foundation`. Think the `HttpKernel` is a little bit unexpected place for the Signer. Would have expected it to be part of the routing component or http foundation component. Atleast I would have searched there for its documentation.

I did stumble over the [UrlHelper](https://symfony.com/doc/current/components/http_foundation.html#generating-relative-and-absolute-urls) and would think the **UriSigner** could be side by side to it.

What do you think?

### Example

```diff
-Symfony\Component\HttpKernel\UriSigner;
+Symfony\Component\HttpFoundation\UriSigner;
```

Reference: [PR Introduction of UriSigner](https://github.com/symfony/symfony/pull/6459)

### PR DESCRIPTION
**Pull request by alexander-schranz**
| Q             | A
| ------------- | ---
| Branch?       | 6.4
| Bug fix?      | no
| New feature?  | yes
| Deprecations? | yes
| Tickets       | Fix #50384
| License       | MIT
| Doc PR        | TODO

Move UriSigner from HttpKernel to HttpFoundation package as discussed in https://github.com/symfony/symfony/issues/50384.

### CODE REVIEW

File: src/Symfony/Component/HttpFoundation/CHANGELOG.md [hunk 1]
```diff
@@ -4,6 +4,7 @@ CHANGELOG
 6.3
 ---
 
+ * Add `UriSigner` from HttpKernel
```
Comment (nicolas-grekas) [inline src/Symfony/Component/HttpFoundation/CHANGELOG.md]:
```suggestion
 * Add `UriSigner`
```

File: src/Symfony/Component/HttpKernel/UriSigner.php [hunk 2]
```diff
@@ -12,11 +12,14 @@
...
 
 use Symfony\Component\HttpFoundation\Request;
+use Symfony\Component\HttpFoundation\UriSigner as HttpFoundationUriSigner;
 
 /**
  * Signs URIs.
  *
  * @author Fabien Potencier <fabien@symfony.com>
+ *
+ * @deprecated since Symfony 6.3, use {@link HttpFoundationUriSigner} instead
```
Comment (nicolas-grekas) [inline src/Symfony/Component/HttpKernel/UriSigner.php]: 6.3

File: src/Symfony/Component/HttpFoundation/CHANGELOG.md [hunk 3]
```diff
@@ -4,6 +4,7 @@ CHANGELOG
 6.3
```
Comment (nicolas-grekas) [inline src/Symfony/Component/HttpFoundation/CHANGELOG.md:11]: should be moved to a new 6.4 section

File: src/Symfony/Bridge/Twig/Tests/Extension/HttpKernelExtensionTest.php [hunk 4]
```diff
@@ -18,10 +18,10 @@
 use Symfony\Component\HttpFoundation\Request;
 use Symfony\Component\HttpFoundation\RequestStack;
 use Symfony\Component\HttpFoundation\Response;
+use Symfony\Component\HttpFoundation\UriSigner;
```
Comment (nicolas-grekas) [inline src/Symfony/Bridge/Twig/Tests/Extension/HttpKernelExtensionTest.php:21]: this requires a version bump in composer.json

File: src/Symfony/Bundle/FrameworkBundle/Resources/config/services.php [hunk 5]
```diff
@@ -35,6 +35,7 @@
 use Symfony\Component\HttpFoundation\RequestStack;
 use Symfony\Component\HttpFoundation\Response;
 use Symfony\Component\HttpFoundation\Session\SessionInterface;
+use Symfony\Component\HttpFoundation\UriSigner;
```
Comment (nicolas-grekas) [inline src/Symfony/Bundle/FrameworkBundle/Resources/config/services.php:38]: this requires a version bump in composer.json

File: src/Symfony/Component/HttpFoundation/Tests/UriSignerTest.php [hunk 6]
```diff
@@ -9,11 +9,11 @@
...
 
-namespace Symfony\Component\HttpKernel\Tests;
+namespace Symfony\Component\HttpFoundation\Tests;
 
 use PHPUnit\Framework\TestCase;
 use Symfony\Component\HttpFoundation\Request;
-use Symfony\Component\HttpKernel\UriSigner;
+use Symfony\Component\HttpFoundation\UriSigner;
 
 class UriSignerTest extends TestCase
```
Comment (nicolas-grekas) [inline src/Symfony/Component/HttpFoundation/Tests/UriSignerTest.php:18]: should be duplicated, not moved
the legacy one should be marked `@group legacy`

File: src/Symfony/Component/HttpFoundation/UriSigner.php [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+namespace Symfony\Component\HttpFoundation;
+
+use Symfony\Component\HttpKernel\UriSigner as HttpKernelUriSigner;
+
+/**
+ * Signs URIs.
+ *
+ * @author Fabien Potencier <fabien@symfony.com>
+ */
+class UriSigner extends HttpKernelUriSigner
```
Comment (nicolas-grekas) [inline src/Symfony/Component/HttpFoundation/UriSigner.php]: we should rather duplicate the code IMHO, we can't make http-kernel a dep of http-foundation

File: src/Symfony/Component/HttpKernel/Tests/EventListener/FragmentListenerTest.php [hunk 8]
```diff
@@ -13,11 +13,11 @@
 
 use PHPUnit\Framework\TestCase;
 use Symfony\Component\HttpFoundation\Request;
+use Symfony\Component\HttpFoundation\UriSigner;
```
Comment (nicolas-grekas) [inline src/Symfony/Component/HttpKernel/Tests/EventListener/FragmentListenerTest.php:16]: requires a version bump

File: src/Symfony/Component/HttpKernel/UriSigner.php.orig [hunk 9]
```diff
@@ -0,0 +1,116 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
```
Comment (nicolas-grekas) [inline src/Symfony/Component/HttpKernel/UriSigner.php.orig]: bad push :)

File: src/Symfony/Component/HttpFoundation/UriSigner.php (hunk 7 above)
Comment (alexander-schranz) [reply src/Symfony/Component/HttpFoundation/UriSigner.php]: Would work to extend the HttpKernel UriSigner from the HttpFoundation UriSigner.

I first did have that but now I don't remember why I did switch it 🙈

Comment (alexander-schranz) [reply src/Symfony/Component/HttpFoundation/UriSigner.php]: I remember now the only problem then would be that the `uri_signer` service need to link to the old http kernel class else it would be a bc break. So only the `Symfony\Component\HttpFoundation\UriSigner` could be wired to the none deprecated instance of UriSigner.

The good one is if we extend the `Symfony\Component\HttpKernel\UriSigner` could still be given to service already switched to the new `Symfony\Component\HttpFoundation\UriSigner` (ESI, ... could then already be switched without bc break).

File: src/Symfony/Bundle/FrameworkBundle/Resources/config/services.php [hunk 10]
```diff
@@ -47,7 +48,6 @@
 use Symfony\Component\HttpKernel\HttpKernelInterface;
 use Symfony\Component\HttpKernel\KernelEvents;
 use Symfony\Component\HttpKernel\KernelInterface;
-use Symfony\Component\HttpKernel\UriSigner;
```
Comment (HypeMC) [inline src/Symfony/Bundle/FrameworkBundle/Resources/config/services.php:50]: This removes the current service alias which is a BC break, it should be deprecated instead, see #46602

File: src/Symfony/Component/HttpFoundation/UriSigner.php (hunk 7 above)
Comment (wouterj) [reply src/Symfony/Component/HttpFoundation/UriSigner.php]: As the class is moved without any logic change, you can use `class_alias`.

As an example, see the Messenger transport classes that were moved from [`Transport`](https://github.com/symfony/symfony/blob/5.4/src/Symfony/Component/Messenger/Transport/Doctrine/DoctrineTransport.php) to [Bridges](https://github.com/symfony/symfony/blob/5.4/src/Symfony/Component/Messenger/Bridge/Doctrine/Transport/DoctrineTransport.php#L133-L135).

Also note that we should not trigger a deprecation when it's instantiated, but when it's loaded (this is implemented by the above example).

Comment (alexander-schranz) [reply src/Symfony/Component/HttpFoundation/UriSigner.php]: @wouterj that is a really great solution 👍

File: src/Symfony/Component/HttpKernel/UriSigner.php (hunk 2 above)
Comment (fabpot) [reply src/Symfony/Component/HttpKernel/UriSigner.php]: 6.4 :)

**Comment by fabpot** [comment]
@alexander-schranz Do you have time to finish this PR?

File: src/Symfony/Bundle/FrameworkBundle/Resources/config/services.php (hunk 10 above)
Comment (alexander-schranz) [reply src/Symfony/Bundle/FrameworkBundle/Resources/config/services.php:50]: do we still require the alias when `Symfony\Component\HttpKernel\UriSigner` is an class alias to `Symfony\Component\HttpFoundation\UriSigner`?

**Comment by alexander-schranz** [comment]
The fabbot seems to return a false positive currently:

Bildschirmfoto 2023-08-04 um 12 52 13

File: src/Symfony/Bundle/FrameworkBundle/Resources/config/services.php (hunk 10 above)
Comment (alexander-schranz) [reply src/Symfony/Bundle/FrameworkBundle/Resources/config/services.php:50]: I readded the alias as ReflectionParameter would return from as `->getType()->__toString()` still the old class and I'm not sure if that kind of reflection is used for autowiring.

**Comment by alexander-schranz** [comment]
If I see it correctly all failing tests are related to current 6.4 branch and not the pull request. So I hope I tackled all parts.

File: src/Symfony/Bundle/FrameworkBundle/Resources/config/services.php (hunk 10 above)
Comment (HypeMC) [reply src/Symfony/Bundle/FrameworkBundle/Resources/config/services.php:50]: The alias should be deprecated though.

Comment (alexander-schranz) [reply src/Symfony/Bundle/FrameworkBundle/Resources/config/services.php:50]: I dont see a way to deprecate here as it returns the parent again?

File: src/Symfony/Bundle/FrameworkBundle/Resources/config/services.php [hunk 11]
```diff
@@ -157,6 +158,7 @@ class_exists(WorkflowEvents::class) ? WorkflowEvents::ALIASES : []
                 param('kernel.secret'),
             ])
         ->alias(UriSigner::class, 'uri_signer')
+        ->alias(HttpKernelUriSigner::class, 'uri_signer')
```
Comment (wouterj) [inline src/Symfony/Bundle/FrameworkBundle/Resources/config/services.php:161]:
```suggestion
        ->alias(HttpKernelUriSigner::class, 'uri_signer')
            ->deprecate('symfony/framework-bundle', '6.4', 'The "%alias_id%" alias is deprecated, use "'.UriSigner::class.'" instead.')
```

File: src/Symfony/Bundle/FrameworkBundle/Resources/config/services.php (hunk 5 above)
Comment (wouterj) [reply src/Symfony/Bundle/FrameworkBundle/Resources/config/services.php:38]: Seems like you forgot to update FrameworkBundle's `composer.json`

File: src/Symfony/Component/HttpFoundation/CHANGELOG.md [hunk 12]
```diff
@@ -4,6 +4,7 @@ CHANGELOG
 6.4
 ---
 
+ * Add `UriSigner`
```
Comment (wouterj) [inline src/Symfony/Component/HttpFoundation/CHANGELOG.md]:
```suggestion
 * Add `UriSigner` from the HttpKernel component
```

File: src/Symfony/Component/HttpFoundation/Tests/UriSignerTest.php (hunk 6 above)
Comment (wouterj) [reply src/Symfony/Component/HttpFoundation/Tests/UriSignerTest.php:18]: I see this is now marked as legacy, but it still is moved. The file should exist in both HttpKernel and HttpFoundation, with the one in HttpKernel having `@group legacy`

**Review by wouterj — CHANGES_REQUESTED** [review]

File: src/Symfony/Component/HttpFoundation/CHANGELOG.md (hunk 12 above)
Comment (wouterj) [reply src/Symfony/Component/HttpFoundation/CHANGELOG.md]: We also need an entry to HttpKernel's CHANGELOG file (saying it was deprecated, use the one in HttpFoundation instead) and in the `UPGRADE-6.4.md` file in the root of the project (for both components)

File: src/Symfony/Bundle/FrameworkBundle/Resources/config/services.php (hunk 5 above)
Comment (alexander-schranz) [reply src/Symfony/Bundle/FrameworkBundle/Resources/config/services.php:38]: was already on 6.4: https://github.com/symfony/symfony/blob/614165670d40b1281a747723677ae2655cb6962c/src/Symfony/Bundle/FrameworkBundle/composer.json#L28

**Comment by alexander-schranz** [comment]
@wouterj thx for the review, I hope I didn't miss something but it should be all tackled.

File: UPGRADE-6.4.md [hunk 13]
```diff
@@ -89,8 +89,11 @@ HttpFoundation
 HttpKernel
 ----------
 
+ * Deprecate `UriSigner`, use `HttpFoundation` `UriSigner` instead
  * [BC break] `BundleInterface` no longer extends `ContainerAwareInterface`
  * [BC break] Add native return types to `TraceableEventDispatcher` and to `MergeExtensionConfigurationPass`
+ * `BundleInterface` no longer extends `ContainerAwareInterface`
+ * Add native return types to `TraceableEventDispatcher` and to `MergeExtensionConfigurationPass`
```
Comment (wouterj) [inline UPGRADE-6.4.md]: Hmm, these 2 seems unrelated. I think maybe something went wrong resolving the merge conflict?

File: src/Symfony/Component/HttpKernel/Tests/UriSignerTest.php [hunk 14]
```diff
@@ -9,12 +9,15 @@
  * file that was distributed with this source code.
  */
 
-namespace Symfony\Component\HttpKernel\Tests;
+namespace Symfony\Component\HttpFoundation\Tests;
```
Comment (wouterj) [inline src/Symfony/Component/HttpKernel/Tests/UriSignerTest.php]:
```suggestion
namespace Symfony\Component\HttpKernel\Tests;
```

File: src/Symfony/Component/HttpKernel/Tests/UriSignerTest.php [hunk 15]
```diff
@@ -9,12 +9,15 @@
...
 
 use PHPUnit\Framework\TestCase;
 use Symfony\Component\HttpFoundation\Request;
-use Symfony\Component\HttpKernel\UriSigner;
+use Symfony\Component\HttpFoundation\UriSigner;
```
Comment (wouterj) [inline src/Symfony/Component/HttpKernel/Tests/UriSignerTest.php]:
```suggestion
use Symfony\Component\HttpKernel\UriSigner;
```

**Review by wouterj — APPROVED** [review]
Thanks, just a few things were missed. But other than that, this looks good to go from my side!

**Comment by fabpot** [comment]
@alexander-schranz Do you have time to finish this PR?

**Comment by wouterj** [comment]
If not, please tell us and we (I) will force push the final tweaks to this PR so we can get it in 6.4/7.0 🙂

**Comment by alexander-schranz** [comment]
@wouterj currently on vacation feel free to force push. Else I can finish it also in two weeks after my vacation.

**Comment by wouterj** [comment]
Enjoy your holiday!

This PR should be ready now.

**Review by fabpot — APPROVED** [review]

**Comment by fabpot** [comment]
Thank you @alexander-schranz.
